### ISSUE TITLE
Rollbacks

### ISSUE THREAD

**Issue by swissspidy** [obenland/shiny-updates#292]
See #288.

From [this comment on trac](https://core.trac.wordpress.org/ticket/31532#comment:8)

> We also need to take care of updates and not allowing updates to a version that contains a fatal error. I would love for us to store the current version and restore that version if an update fails, but that may be out of scope for now.

Sounds like an exciting challenge.

**Comment by anttiviljami** [comment]
I suggest we give the user the possibility to manually roll back a recent update.

We could show a dismissable notification like this for a short period of time after the update: 

>  You've recently updated Shiny Updates to version 3. If you see something went wrong, you can always **roll back**.

A manual roll back to recent version button would also be nice.

**Comment by swissspidy** [comment]
Manual rollbacks sounds like a no-no to me. What if a plugin ran some database upgrade routine? After downgrading again things could break horribly.

That's why you should not downgrade your WordPress site, nor your iPhone's operating system or anything else.

**Comment by anttiviljami** [comment]
> What if a plugin ran some database upgrade routine? After downgrading again things could break horribly.

That's why we already test for fatal errors during installs and upgrades. In that case, the rollback would just fail gracefully and leave the plugin deactivated.

**Comment by afragen** [comment]
You can use the WP-Rollback plugin in the repo for this.

**Comment by anttiviljami** [comment]
@afragen i'm proposing adding the rollback feature to core. The plugin seems great btw!

**Comment by afragen** [comment]
I think I'm with @swissspidy. The potential badness likely outweighs any benefit.

**Comment by anttiviljami** [comment]
@afragen: what sort of potential badbess are you talking about? I'm sure I must be missing something here

**Comment by afragen** [comment]
Well are we referring to a general rollback to any version or just rollback to last version because we had issues updating?

If it's rollback to any version, what happens if someone unknowingly rolls back to a version with a security issue?

If we're just talking about the last version if there's an update error then that seems like a good idea. Can still have issues if database updates happen that aren't reverted.

**Comment by afragen** [comment]
@anttiviljami I think your wp-safe-updates should be its own feature project :+1:

**Comment by anttiviljami** [comment]
The mechanism I'm suggesting for rollbacks is that before upgrading to the new version, instead of clearing the old plugin directory by deleting all the files, we would move the old version to another location where it'll be stored for a few days, until it looks like there's no need for a rollback.

Then the user could restore the plugin from that location if they feel something is not right with the upgrade.

Rolling back to just any version wouldn't work very well, since a lot of plugin vendors don't keep old versions in the plugin directory svn.

**Comment by anttiviljami** [comment]
> @anttiviljami I think your wp-safe-updates should be its own feature project 👍

Awesome that you think so. :) Feel free to comment on the issue at! https://core.trac.wordpress.org/ticket/37301

**Comment by swissspidy** [comment]
As long as there's a lack of education and standards for plugin developers to follow, there will be edge cases where things could horribly break when downgrading and upgrading again.

The average user has no idea that this can happen. Giving them a sledge hammer to break their site is pretty bad. In my opinion, this wouldn't be in line with the [WordPress philosophy](https://wordpress.org/about/philosophy/).

----

Long story short: Let's do automatic rollbacks as intended in the original ticket for now. We can always iterate from there. I suggest talking to the developers of the WP-Rollback plugin and opening a ticket on Trac for manual rollbacks if necessary. That's the better place for that :-)

I'm looking forward to getting more feedback from others on this topic, which is important to bring this to the next level.

**Comment by afragen** [comment]
@DevinWalker wondering if you want to chime in?

**Comment by DevinWalker** [comment]
I agree that rollbacks should not be in core due to the incompatibilities with some plugins not using tags properly and complications with upgrade routines. We created WP Rollback to use in case of emergencies or for testing purposes. When we proposed the idea to include it into core to some of the core devs we got a pretty sturdy no.

**Comment by anttiviljami** [comment]
@DevinWalker:

> some plugins not using tags properly

I say leave tags and versions out of it. Just move the pre-update version away for storage before replacing it with the new version and allow the user to go back to the old version if they wish to do so. The user wouldn't remember what version they were on before breaking the update anyway.

> complications with upgrade routines

Yes, rollbacks for some plugins may completely fail due to database incompatibilities between versions. But then again we already have the means to check for fatal errors on plugin activation so this wouldn't really be an issue. The failed rollback would just leave the old version of the plugin deactivated due to the errors.

And if the issue with the rollback is non-fatal, the user can just upgrade back to the new version.

Can we please have some actual discussion about the model I'm suggesting here instead of just saying manual rollbacks are bad for the sake of them being rollbacks? Nobody has yet to give me an example where the manual rollback would potentially break the site more than it already was broken after the upgrade. 

Thanks. :)
